redshift/redshift.go

```diff
@@ -19,59 +17,55 @@ type dbExecCloser interface {
-	awsAccessKeyID     = env.MustGet("AWS_ACCESS_KEY_ID")
-	awsSecretAccessKey = env.MustGet("AWS_SECRET_ACCESS_KEY")
-	host               = flag.String("redshifthost", "", "Address of the redshift host")
-	port               = flag.Int("redshiftport", 0, "Address of the redshift host")
-	db                 = flag.String("redshiftdatabase", "", "Redshift database to connect to")
-	user               = flag.String("redshiftuser", "", "Redshift user to connect as")
-	pwd                = flag.String("redshiftpassword", "", "Password for the redshift user")
-	timeout            = flag.Duration("redshiftconnecttimeout", 10*time.Second,
-		"Timeout while connecting to Redshift. Defaults to 10 seconds.")
 )
 
-// NewRedshift returns a pointer to a new redshift object using configuration values set in the
-// flags.
-func NewRedshift() (*Redshift, error) {
+// NewRedshift returns a pointer to a new redshift object using configuration values passed in
+// on instantiation and the AWS env vars we assume exist
+// Don't need to pass s3 info unless doing a COPY operation
+func NewRedshift(host, port, db, user, pwd string, timeout int, s3Info S3Info) (*Redshift, error) {
 	flag.Parse()
-	source := fmt.Sprintf("host=%s port=%d dbname=%s connect_timeout=%d", *host, *port, *db, int(timeout.Seconds()))
-	log.Println("Connecting to Reshift Source: ", source)
-	source += fmt.Sprintf(" user=%s password=%s", *user, *pwd)
+	source := fmt.Sprintf("host=%s port=%d dbname=%s connect_timeout=%d", host, port, db, timeout)
+	log.Println("Connecting to Redshift Source: ", source)
+	source += fmt.Sprintf(" user=%s password=%s", user, pwd)
 	sqldb, err := sql.Open("postgres", source)
 	if err != nil {
 		return nil, err
 	}
-	return &Redshift{sqldb, awsAccessKeyID, awsSecretAccessKey}, nil
+	return &Redshift{sqldb, s3Info}, nil
 }
 
-func (r *Redshift) logAndExec(cmd string, creds bool) (sql.Result, error) {
+func (r *Redshift) logAndExec(cmd string) (sql.Result, error) {
 	log.Print("Executing Redshift command: ", cmd)
-	if creds {
-		cmd += fmt.Sprintf(" CREDENTIALS '%s=%s;%s=%s'", "aws_access_key_id", r.accessID, "aws_secret_access_key", r.secretKey)
-	}
 	return r.Exec(cmd)
 }
 
 // CopyJSONDataFromS3 copies JSON data present in an S3 file into a redshift table.
-func (r *Redshift) CopyJSONDataFromS3(schema, table, file, jsonpathsFile, awsRegion string) error {
+func (r *Redshift) CopyJSONDataFromS3(schema, table, file, jsonpathsFile string) error {
+	creds := fmt.Sprintf(`CREDENTIALS 'aws_access_key_id=%s;aws_secret_access_key=%s'`, r.s3Info.AccessID, r.s3Info.SecretKey)
 	copyCmd := fmt.Sprintf(
-		`COPY "%s"."%s" FROM '%s' WITH json '%s' region '%s' timeformat 'epochsecs' COMPUPDATE ON`,
-		schema, table, file, jsonpathsFile, awsRegion)
-	_, err := r.logAndExec(copyCmd, true)
+		`COPY "%s"."%s" FROM '%s' WITH json '%s' region '%s' timeformat 'epochsecs' COMPUPDATE ON %s`,
+		schema, table, file, jsonpathsFile, r.s3Info.Region, creds,
+	)
+	_, err := r.logAndExec(copyCmd)
```

Comment:
hmm, this is a huge pain, I might take this time to go back to calling the copy functions directly, but start using go's actual transaction library instead of calling begin and end transaction as well, which solves some of the pain that returning the sql was solving...